rs/messaging/tests/memory_tests.rs

```diff
@@ -0,0 +1,500 @@
+/// back to 0 after all in-flight messages have been dealt with.
+fn check_guaranteed_response_message_memory_limits_are_respected_impl(
+    seeds: &[u64],
+    max_payload_bytes: u32,
+    calls_per_round: u32,
+    reply_weight: u32,
+    call_weight: u32,
+) -> Result<(), (String, DebugInfo)> {
+    // The number of rounds to execute while chatter is on.
+    const CHATTER_PHASE_ROUND_COUNT: u64 = 30;
+    // The maximum number of rounds to execute after chatter is turned off. It it takes more than
+    // this number of rounds until there are no more hanging calls, the test fails.
+    const SHUTDOWN_PHASE_MAX_ROUNDS: u64 = 300;
+    // The amount of memory available for guaranteed response message memory on `local_env`.
+    const LOCAL_MESSAGE_MEMORY_CAPACITY: u64 = 100 * MB;
+    // The amount of memory available for guaranteed response message memory on `remote_env`.
+    const REMOTE_MESSAGE_MEMORY_CAPACITY: u64 = 50 * MB;
+
+    let fixture = Fixture::new(FixtureConfig {
+        local_canisters_count: 2,
+        local_max_instructions_per_round: 100_000_000,
+        local_message_memory_capacity: LOCAL_MESSAGE_MEMORY_CAPACITY,
+        remote_canisters_count: 1,
+        remote_max_instructions_per_round: 100_000_000,
+        remote_message_memory_capacity: REMOTE_MESSAGE_MEMORY_CAPACITY,
+    });
+
+    let config = CanisterConfig::try_new(
+        fixture.canisters(),   // receivers
+        0..=max_payload_bytes, // call_bytes
+        0..=max_payload_bytes, // reply_bytes
+        0..=0,                 // instructions_count
+    )
+    .unwrap();
+
+    // Send configs to canisters, seed the rng.
+    for (index, canister) in fixture.canisters().into_iter().enumerate() {
+        fixture.replace_config(canister, config.clone()).unwrap();
+        fixture.seed_rng(canister, seeds[index]);
+        fixture
+            .replace_reply_weight(canister, reply_weight)
+            .unwrap();
+        fixture.replace_call_weight(canister, call_weight).unwrap();
+    }
+
+    // Start chatter on all canisters.
+    fixture.start_chatter(calls_per_round).unwrap();
+
+    // Build up backlog and keep up chatter for while.
+    for _ in 0..CHATTER_PHASE_ROUND_COUNT {
+        fixture.tick();
+
+        // Check message memory limits are respected.
+        let (local_memory, remote_memory) = fixture.guaranteed_response_message_memory_taken();
+        if local_memory > LOCAL_MESSAGE_MEMORY_CAPACITY.into() {
+            return fixture.debug_info("Chatter: local memory exceeds limit");
+        }
+        if remote_memory > REMOTE_MESSAGE_MEMORY_CAPACITY.into() {
+            return fixture.debug_info("Chatter: remote memory exceeds limit");
+        }
+    }
+
+    // Stop chatter on all canisters; stop additional downstream calls.
+    fixture.stop_chatter().unwrap();
+    for canister in fixture.canisters() {
+        fixture.replace_call_weight(canister, 0).unwrap();
+    }
```

Comment:
Why do we do this? Weren't the weights themselves supposed to result in a long tail distribution? (Or I guess this is why they are both `0..=2`.)

Regardless, I would make this part of `stop_chatter()` if the chatter won't stop without also doing this.

Reply:
It depends on what kind of weights you pick. `stop_chatter()` makes the heartbeat produce 0 messages. For a completely random call tree I'm not sure how to compute it, but for just a chain of downstream calls the expected depth is:
```
x = call_weight / (reply_weight + call_weight); expected_depth = x / (1 - x).
```
So it diverges only for `reply_weight == 0` (this doesn't take into account synchronous rejections though).

Anyways, since downstream calls are made randomly after the subtree was 'planted' by the heartbeat, the only way to stop this immediately is to set the call weight to 0.

I didn't put this into `stop_chatter()` because if you later call `start_chatter()` it won't do the same as it did before because the call weights are now 0. I put it into `stop_chatter()` and mentioned it in the doc comment.